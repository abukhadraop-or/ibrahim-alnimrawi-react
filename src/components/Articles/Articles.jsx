import LikeButton from 'components/shared/LikeButton/LikeButton';
import Pagination from 'components/shared/Pagination/Pagination';
import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import { getArticles } from 'services/fake-articles-service';
import { formatDate } from 'utils/format-date';
import {
  Article,
  ArticleContainer,
  ArticleDate,
  ArticleDescription,
  ArticleHeader,
  ArticleReadMore,
  ArticleSubContainer,
  ArticleTitle,
  ArticleUser,
  ArticleUserImage,
  Container,
  NotFound,
} from './articles.style';

/**
 * Render a <Articles> component
 *
 * @param {Object} props     The props object.
 * @param {String} props.tag The name of the current tag.
 *
 * @return {JSX.Element}
 */
const Articles = ({ tag }) => {
  const [articles, setArticles] = useState([]);
  const [page, setPage] = useState(1);
  const [numberOfArticles, setNumberOfArticles] = useState(0);

  /**
   * Set the page to the selected page.
   *
   * @param {Number} selectedPage The number of selected page.
   */
  const onPageChange = (selectedPage) => setPage(selectedPage);

  // handle initial state.
  useEffect(() => {
    const data = async () => {
      const response = await getArticles(tag, page, 10);
      setArticles(response.articles);
      setNumberOfArticles(response.articlesCount);
    };
    data();
  }, [page]);

  // handle filtering.
  useEffect(() => {
    setPage(1);
    const data = async () => {
      const response = await getArticles(tag, page, 10);
      setArticles(response.articles);
      setNumberOfArticles(response.articlesCount);
    };
    data();
  }, [tag]);

  return (
    <Container>
      {!numberOfArticles && <NotFound>No articles are here... yet.</NotFound>}
      {!!numberOfArticles && (
        <Article>
          {numberOfArticles > 0 &&
            articles.map((article) => (
              <ArticleContainer key={article.slug}>
                <ArticleHeader>
                  <ArticleUserImage ImgSrc={article.author.image} />
                  <ArticleSubContainer>
                    <ArticleUser>{article.author.username}</ArticleUser>
                    <ArticleDate>{formatDate(article.createdAt)}</ArticleDate>
                  </ArticleSubContainer>
                  <LikeButton count={article.favoritesCount} />
                </ArticleHeader>
                <ArticleTitle>{article.title}</ArticleTitle>
                <ArticleDescription>{article.description}</ArticleDescription>
                <ArticleReadMore>Read more...</ArticleReadMore>
              </ArticleContainer>
            ))}
        </Article>
      )}

      <Pagination
        onPageChange={onPageChange}
        currentPage={page}
        pageSize={10}
        itemsCount={numberOfArticles}
      />
    </Container>
  );
};

export default Articles;

Articles.propTypes = {
  tag: PropTypes.string,
};

Articles.defaultProps = {
  tag: '',
};
