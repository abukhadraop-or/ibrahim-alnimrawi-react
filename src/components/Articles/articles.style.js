import colors from 'style/colors';
import mediaQueries from 'style/media-queries';
import styled from 'styled-components';

export const ArticleHeader = styled.div`
  display: flex;
  flex-direction: row;
`;

export const ArticleSubContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

export const ArticleUser = styled.div`
  color: ${colors.moderateLimeGreen};
  cursor: pointer;

  &:hover {
    color: ${colors.darkCyan_limeGreen};
    text-decoration: underline;
  }
`;

export const ArticleDate = styled.div`
  color: ${colors.gray};
  font-size: 0.8rem;
`;

/* istanbul ignore next */
export const ArticleUserImage = styled.div`
  background-image: url(${(props) => props.ImgSrc});
  background-size: cover;
  border-radius: 50%;
  cursor: pointer;
  height: 2rem;
  margin-right: 0.5rem;
  width: 2rem;
`;

export const ArticleTitle = styled.h1`
  cursor: pointer;
  font-size: 1.5rem;
  font-weight: 600;
`;

export const ArticleBody = styled.div`
  font-size: 1rem;
  font-weight: 300;
`;

export const Article = styled.div`
  border-top: 0.063rem solid ${colors.Mercury};
  flex-direction: column;
  margin: 0 5rem;
  padding: 1.5rem 0;
`;

export const ArticleDescription = styled.div`
  color: ${colors.darkGrayishBlue};
  cursor: pointer;
  margin-bottom: 0.9rem;
`;

export const ArticleReadMore = styled.div`
  color: ${colors.darkGrayishBlue};
  cursor: pointer;
  font-size: 0.8rem;
  font-weight: 300;
`;

export const ArticleSection = styled.div`
  display: flex;
  justify-content: space-between;

  @media (max-width: ${mediaQueries.mobile}) {
    flex-direction: column;
  }
`;

export const ArticleContainer = styled.div`
  margin-bottom: 3rem;
`;

export const NotFound = styled.div`
  border: none;
  color: ${colors.veryDarkGrayishBlue3};
  margin: 1rem 5rem;
`;

export const Container = styled.div`
  display: flex;
  flex-direction: column;
`;
