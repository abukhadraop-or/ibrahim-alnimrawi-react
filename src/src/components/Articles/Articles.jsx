import LikeButton from "components/Shared/LikeButton/LikeButton";
import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import {
  getArticlesNumber,
  paginateSlice,
} from "../../services/fakeArticlesService";
import Pagination from "../Shared/Pagination/Pagination";
import {
  Article,
  ArticleContainer,
  ArticleDate,
  ArticleDescription,
  ArticleHeader,
  ArticleReadMore,
  ArticleSubContainer,
  ArticleTitle,
  ArticleUser,
  ArticleUserImage,
  Container,
  NotFound,
} from "./articles.style";

/**
 * Handle date format.
 *
 * @param {string} originalDate Article created date.
 *
 * @return {string}             Date in the right format.
 */
const handleDate = (originalDate) => {
  const date = new Date(Date.parse(originalDate));
  return date.toDateString();
};

/**
 * Render a <Articles> component
 *
 * @param {String} props.tag The name of the current tag.
 *
 * @return {JSX.Element}
 */
const Articles = ({ tag }) => {
  const [articles, setArticles] = useState([]);
  const [page, setPage] = useState(1);
  const [numberOfArticles, setNumberOfArticles] = useState(0);
  const handlePageChange = (page) => {
    setPage(page);
    setArticles(paginateSlice(page, 10));
  };

  useEffect(() => {
    const data = async () => {
      let response = await paginateSlice(tag, page, 10);
      setArticles(response);
      setNumberOfArticles(getArticlesNumber);
    };

    data();
  }, [page, tag]);

  useEffect(() => {
    if (!!tag) {
      const data = paginateSlice(tag, page, 10);
      setNumberOfArticles(data[0].length);

      setArticles(data[0]);
    }
  }, [tag, numberOfArticles, page]);
  return (
    <Container>
      {!numberOfArticles && <NotFound>No articles are here... yet.</NotFound>}
      {!!numberOfArticles && (
        <Article>
          {articles.length > 0 &&
            articles.map((article) => (
              <ArticleContainer key={article.slug}>
                <ArticleHeader>
                  <ArticleUserImage ImgSrc={article?.author?.image} />
                  <ArticleSubContainer>
                    <ArticleUser>{article?.author?.username}</ArticleUser>
                    <ArticleDate>{handleDate(article?.createdAt)}</ArticleDate>
                  </ArticleSubContainer>
                  <LikeButton count={article?.favoritesCount}></LikeButton>
                </ArticleHeader>
                <ArticleTitle>{article.title}</ArticleTitle>
                <ArticleDescription>{article.description}</ArticleDescription>
                <ArticleReadMore>Read more...</ArticleReadMore>
              </ArticleContainer>
            ))}
        </Article>
      )}

      <Pagination
        onPageChange={handlePageChange}
        currentPage={page}
        pageSize={10}
        itemsCount={numberOfArticles}
      />
    </Container>
  );
};

export default Articles;

Articles.propTypes = {
  tag: PropTypes.string,
};
