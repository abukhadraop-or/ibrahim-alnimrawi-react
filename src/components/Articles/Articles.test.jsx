import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme from 'enzyme';
import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import Articles from './Articles';

Enzyme.configure({ adapter: new Adapter() });
describe('<Home />', () => {
  test('it match the snapshot be', () => {
    const tree = renderer.create(<Articles />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
