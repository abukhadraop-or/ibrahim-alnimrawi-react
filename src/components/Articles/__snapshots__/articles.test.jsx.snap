// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> it match the snapshot be 1`] = `
.c1 {
  border: none;
  color: #2F3236;
  margin: 1rem 5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    No articles are here... yet.
  </div>
</div>
`;
